File: substrate/floating_instance_tray_compose.py
"""Floating deep-research instance tray compose (pure).

pack_dispatched, merge_executed, live_dispatched always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal

TrayAction = Literal[
    "none",
    "fullscreen_one",
    "collective_pack",
    "cohesive_prompt",
    "draft_merge_one",
    "full_merge_one",
]
VALID_ACTIONS = frozenset(
    (
        "none",
        "fullscreen_one",
        "collective_pack",
        "cohesive_prompt",
        "draft_merge_one",
        "full_merge_one",
    )
)
VALID_STATUS = frozenset(("proposed", "open", "completed", "closed"))


class FloatingInstanceTrayComposeError(ValueError):
    """Fail-closed validation for floating instance tray."""


@dataclass(frozen=True)
class FloatingInstanceTrayCompose:
    parent_asset_id: str
    member_count: int
    selected_count: int
    selected_instance_ids: tuple[str, ...]
    action: str
    tray_ready: bool
    pack_dispatched: bool
    merge_executed: bool
    live_dispatched: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "parent_asset_id": self.parent_asset_id,
            "member_count": self.member_count,
            "selected_count": self.selected_count,
            "selected_instance_ids": list(self.selected_instance_ids),
            "action": self.action,
            "tray_ready": self.tray_ready,
            "pack_dispatched": False,
            "merge_executed": False,
            "live_dispatched": False,
            "notes": list(self.notes),
            "authority": "floating_instance_tray_compose_advisory",
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise FloatingInstanceTrayComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_floating_instance_tray(
    *,
    parent_asset_id: object,
    members: object,
    selected_instance_ids: object,
    action: object,
    operator_ack: object,
) -> FloatingInstanceTrayCompose:
    """Compose multi-instance tray readiness. Never dispatches or merges."""
    if not isinstance(operator_ack, bool):
        raise FloatingInstanceTrayComposeError(
            "operator_ack must be an explicit boolean"
        )
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id")
    if not isinstance(members, list) or len(members) == 0:
        raise FloatingInstanceTrayComposeError(
            "members must be a non-empty array"
        )
    if not isinstance(selected_instance_ids, list):
        raise FloatingInstanceTrayComposeError(
            "selected_instance_ids must be an array"
        )
    if action not in VALID_ACTIONS:
        raise FloatingInstanceTrayComposeError("action invalid")

    notes: list[str] = [
        "pack_dispatched=false — tray is pure selection/intent only",
        "merge_executed=false — no parent merge from tray",
        "live_dispatched=false — no provider dispatch",
    ]

    by_id: dict[str, dict[str, Any]] = {}
    for i, m in enumerate(members):
        if not isinstance(m, dict):
            raise FloatingInstanceTrayComposeError(
                f"members[{i}] must be an object"
            )
        mid = _require_nonempty(m.get("instance_id"), field=f"members[{i}].instance_id")
        p = _require_nonempty(
            m.get("parent_asset_id"), field=f"members[{i}].parent_asset_id"
        )
        if p != parent:
            raise FloatingInstanceTrayComposeError(
                "all members must share parent_asset_id"
            )
        st = m.get("status")
        if st not in VALID_STATUS:
            raise FloatingInstanceTrayComposeError(
                f"members[{i}].status invalid"
            )
        if m.get("live_dispatched") is not None and m.get("live_dispatched") is not False:
            raise FloatingInstanceTrayComposeError(
                f"members[{i}].live_dispatched must be false when set"
            )
        if m.get("merge_executed") is not None and m.get("merge_executed") is not False:
            raise FloatingInstanceTrayComposeError(
                f"members[{i}].merge_executed must be false when set"
            )
        if mid in by_id:
            raise FloatingInstanceTrayComposeError(
                f"duplicate instance_id: {mid}"
            )
        by_id[mid] = m

    selected: list[str] = []
    seen: set[str] = set()
    for i, raw in enumerate(selected_instance_ids):
        sid = _require_nonempty(raw, field=f"selected_instance_ids[{i}]")
        if sid not in by_id:
            raise FloatingInstanceTrayComposeError(
                f"selected_instance_ids[{i}] not in members"
            )
        if sid in seen:
            raise FloatingInstanceTrayComposeError(
                f"duplicate selected_instance_id: {sid}"
            )
        seen.add(sid)
        selected.append(sid)

    member_count = len(by_id)
    selected_count = len(selected)
    notes.append(
        f"member_count={member_count} · selected_count={selected_count}"
    )

    tray_ready = False
    if action == "none":
        notes.append("action=none — no tray action selected")
    elif action in ("fullscreen_one", "draft_merge_one", "full_merge_one"):
        if selected_count != 1:
            notes.append(f"action={action} requires exactly 1 selected instance")
        else:
            m = by_id[selected[0]]
            st = m["status"]
            if st == "closed":
                notes.append(f"action={action} blocked — instance closed")
            elif action == "full_merge_one":
                if st != "completed":
                    notes.append("full_merge_one requires completed instance")
                elif not operator_ack:
                    notes.append("full_merge_one requires operator_ack")
                else:
                    tray_ready = True
                    notes.append("tray_ready=true · full_merge_one intent only")
            elif action == "draft_merge_one":
                if st not in ("proposed", "open", "completed"):
                    notes.append(
                        "draft_merge_one requires proposed|open|completed"
                    )
                else:
                    tray_ready = True
                    notes.append("tray_ready=true · draft_merge_one intent only")
            else:
                tray_ready = True
                notes.append("tray_ready=true · fullscreen_one view intent")
    else:
        if selected_count < 2:
            notes.append(f"action={action} requires ≥2 selected instances")
        else:
            ok = True
            for sid in selected:
                m = by_id[sid]
                if m["status"] == "closed":
                    ok = False
                    notes.append(f"selected {sid} is closed")
                    break
            if ok and not operator_ack:
                notes.append(f"{action} requires operator_ack")
            elif ok:
                tray_ready = True
                notes.append(
                    f"tray_ready=true · {action} multi-select intent only (no dispatch)"
                )

    notes.extend(
        (
            "pack_dispatched=false",
            "merge_executed=false",
            "live_dispatched=false",
        )
    )

    return FloatingInstanceTrayCompose(
        parent_asset_id=parent,
        member_count=member_count,
        selected_count=selected_count,
        selected_instance_ids=tuple(selected),
        action=str(action),
        tray_ready=tray_ready,
        pack_dispatched=False,
        merge_executed=False,
        live_dispatched=False,
        notes=tuple(notes),
        authority="floating_instance_tray_compose_advisory",
    )


__all__ = [
    "FloatingInstanceTrayCompose",
    "FloatingInstanceTrayComposeError",
    "compose_floating_instance_tray",
]

File: apps/reading/src/api/floatingInstanceTrayCompose.ts
/**
 * Floating deep-research instance tray compose (pure).
 *
 * Operator vision: multiple floating deep researches open at once; select
 * one or many for fullscreen, collective pack, cohesive unit prompt, draft
 * merge, or full merge — without live dispatch.
 *
 * pack_dispatched always false.
 * merge_executed always false.
 * live_dispatched always false.
 */

export type TrayMemberStatus =
  | "proposed"
  | "open"
  | "completed"
  | "closed";

export type TrayAction =
  | "none"
  | "fullscreen_one"
  | "collective_pack"
  | "cohesive_prompt"
  | "draft_merge_one"
  | "full_merge_one";

export interface TrayMember {
  instance_id: string;
  parent_asset_id: string;
  status: TrayMemberStatus;
  view_mode?: string;
  highlight?: string;
  /** Always false when supplied — pure tray rejects live_dispatched true. */
  live_dispatched?: false;
  merge_executed?: false;
}

export interface FloatingInstanceTrayInput {
  parent_asset_id: string;
  members: TrayMember[];
  /** Selected instance ids (subset of members). */
  selected_instance_ids: string[];
  /** Intended tray action for selection. */
  action: TrayAction;
  operator_ack: boolean;
}

export interface FloatingInstanceTrayCompose {
  parent_asset_id: string;
  member_count: number;
  selected_count: number;
  selected_instance_ids: string[];
  action: TrayAction;
  /**
   * True when selection + action gates pass (and operator_ack for merge/pack).
   * Still never dispatches or merges.
   */
  tray_ready: boolean;
  /** Always false — collective pack never dispatches. */
  pack_dispatched: false;
  /** Always false — merge never executes. */
  merge_executed: false;
  /** Always false — no provider dispatch. */
  live_dispatched: false;
  notes: string[];
  authority: "floating_instance_tray_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Compose multi-instance floating tray readiness for operator actions.
 * Never live-dispatches; never merges.
 */
export function composeFloatingInstanceTray(
  input: FloatingInstanceTrayInput,
): FloatingInstanceTrayCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const parent_asset_id = requireNonEmpty(
    input.parent_asset_id,
    "parent_asset_id",
  );
  if (!Array.isArray(input.members) || input.members.length === 0) {
    throw new Error("members must be a non-empty array");
  }
  if (!Array.isArray(input.selected_instance_ids)) {
    throw new Error("selected_instance_ids must be an array");
  }
  const action = input.action;
  if (
    action !== "none" &&
    action !== "fullscreen_one" &&
    action !== "collective_pack" &&
    action !== "cohesive_prompt" &&
    action !== "draft_merge_one" &&
    action !== "full_merge_one"
  ) {
    throw new Error("action invalid");
  }

  const notes: string[] = [
    "pack_dispatched=false — tray is pure selection/intent only",
    "merge_executed=false — no parent merge from tray",
    "live_dispatched=false — no provider dispatch",
  ];

  const byId = new Map<string, TrayMember>();
  for (let i = 0; i < input.members.length; i++) {
    const m = input.members[i];
    if (!m || typeof m !== "object") {
      throw new Error(`members[${i}] must be an object`);
    }
    const id = requireNonEmpty(m.instance_id, `members[${i}].instance_id`);
    const p = requireNonEmpty(
      m.parent_asset_id,
      `members[${i}].parent_asset_id`,
    );
    if (p !== parent_asset_id) {
      throw new Error("all members must share parent_asset_id");
    }
    if (
      m.status !== "proposed" &&
      m.status !== "open" &&
      m.status !== "completed" &&
      m.status !== "closed"
    ) {
      throw new Error(`members[${i}].status invalid`);
    }
    if (m.live_dispatched !== undefined && m.live_dispatched !== false) {
      throw new Error(`members[${i}].live_dispatched must be false when set`);
    }
    if (m.merge_executed !== undefined && m.merge_executed !== false) {
      throw new Error(`members[${i}].merge_executed must be false when set`);
    }
    if (byId.has(id)) {
      throw new Error(`duplicate instance_id: ${id}`);
    }
    byId.set(id, m);
  }

  const selected_instance_ids: string[] = [];
  const seenSel = new Set<string>();
  for (let i = 0; i < input.selected_instance_ids.length; i++) {
    const id = requireNonEmpty(
      input.selected_instance_ids[i],
      `selected_instance_ids[${i}]`,
    );
    if (!byId.has(id)) {
      throw new Error(`selected_instance_ids[${i}] not in members`);
    }
    if (seenSel.has(id)) {
      throw new Error(`duplicate selected_instance_id: ${id}`);
    }
    seenSel.add(id);
    selected_instance_ids.push(id);
  }

  const member_count = byId.size;
  const selected_count = selected_instance_ids.length;
  notes.push(`member_count=${member_count} · selected_count=${selected_count}`);

  let tray_ready = false;
  if (action === "none") {
    tray_ready = false;
    notes.push("action=none — no tray action selected");
  } else if (
    action === "fullscreen_one" ||
    action === "draft_merge_one" ||
    action === "full_merge_one"
  ) {
    if (selected_count !== 1) {
      notes.push(`action=${action} requires exactly 1 selected instance`);
    } else {
      const m = byId.get(selected_instance_ids[0])!;
      if (m.status === "closed") {
        notes.push(`action=${action} blocked — instance closed`);
      } else if (action === "full_merge_one") {
        if (m.status !== "completed") {
          notes.push("full_merge_one requires completed instance");
        } else if (!input.operator_ack) {
          notes.push("full_merge_one requires operator_ack");
        } else {
          tray_ready = true;
          notes.push("tray_ready=true · full_merge_one intent only");
        }
      } else if (action === "draft_merge_one") {
        if (
          m.status !== "proposed" &&
          m.status !== "open" &&
          m.status !== "completed"
        ) {
          notes.push("draft_merge_one requires proposed|open|completed");
        } else {
          tray_ready = true;
          notes.push("tray_ready=true · draft_merge_one intent only");
        }
      } else {
        // fullscreen_one
        tray_ready = true;
        notes.push("tray_ready=true · fullscreen_one view intent");
      }
    }
  } else {
    // collective_pack | cohesive_prompt
    if (selected_count < 2) {
      notes.push(`action=${action} requires ≥2 selected instances`);
    } else {
      let ok = true;
      for (const id of selected_instance_ids) {
        const m = byId.get(id)!;
        if (m.status === "closed") {
          ok = false;
          notes.push(`selected ${id} is closed`);
          break;
        }
        if (
          action === "collective_pack" &&
          m.status !== "open" &&
          m.status !== "completed" &&
          m.status !== "proposed"
        ) {
          ok = false;
          notes.push(`collective_pack rejects status=${m.status} on ${id}`);
          break;
        }
      }
      if (ok && !input.operator_ack) {
        notes.push(`${action} requires operator_ack`);
      } else if (ok) {
        tray_ready = true;
        notes.push(
          `tray_ready=true · ${action} multi-select intent only (no dispatch)`,
        );
      }
    }
  }

  notes.push("pack_dispatched=false");
  notes.push("merge_executed=false");
  notes.push("live_dispatched=false");

  return {
    parent_asset_id,
    member_count,
    selected_count,
    selected_instance_ids,
    action,
    tray_ready,
    pack_dispatched: false,
    merge_executed: false,
    live_dispatched: false,
    notes,
    authority: "floating_instance_tray_compose_advisory",
  };
}

export function formatFloatingInstanceTraySummary(
  c: FloatingInstanceTrayCompose,
): string {
  return (
    `tray_ready=${c.tray_ready} · action=${c.action} · ` +
    `selected=${c.selected_count}/${c.member_count} · ` +
    `pack_dispatched=false · merge_executed=false`
  );
}
